common/djangoapps/student/signals/receivers.py

```diff
@@ -17,3 +23,18 @@ def update_last_login(sender, user, **kwargs):  # pylint: disable=unused-argumen
     if not waffle().is_enabled(PREVENT_AUTH_USER_WRITES):
         user.last_login = timezone.now()
         user.save(update_fields=['last_login'])
+
+
+def on_user_updated(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Check for retired usernames.
+    """
+    # Check only at User creation time and when not raw.
+    if not instance.id and not kwargs['raw']:
+        # Check for username that's too close to retired username format.
```

Comment:
Sorry to bang on this retired format thing again, but the more I think about this the more I wonder if this check makes sense. In the current form this wouldn't catch any users where the format token was at the beginning of the string. It's also very flaky in cases where the number of characters before the format token is small. Ex. "z{}_retired" would block any usernames starting with "z". Maybe we can make some clear requirements where the setting is configured so that people are warned from using format strings that break the checks, or maybe it just makes sense to skip the check given the huge hash space we're looking at.

Reply:
No problem! I'd rather just skip the check altogether. It's of minimal use IMO. Even if it blocks usage of usernames prefixed with `retired_username_`, here's a list of other prefixes that *aren't* blocked:
```
retired__username_<hash>
retired__username__<hash>
retired_username<hash>
retired-username_<hash>
retired-username-<hash>
retired_usernane_<hash>
retire_username_<hash>
retireed_username_<hash>
(Should I go on.... ;) )
```
All of the above could be confusing to an end user, namely a customer support representative who is attempting to determine if a learner has been retired by looking at their username alone.

@macdiesel I believe you may have added this AC to the ticket initially? If so, how do you feel about removing this prefix check?

Reply:
There is always going to be someone who will find a way around the system and create something similar. I do think it will behoove us to enforce some type of blocking. What about a regex that looks for words "retired" and "username" in a string and blocks it's creation no matter how many underscores, spaces, characters etc are between the words?

Reply:
We've made the retired username format configurable via a Django setting - for Open edX installations to use whatever they please. So we're not guaranteed to have either `retired` or `username` present in the format. The only requirement for the format is that it has a `{}` somewhere in it for the hash insertion. We could go down the path of specifying a list of words separately that cause a username to be blocked. But I'd rather we kept things simple and just remove the check altogether. Further thoughts?

Reply:
What about a django setting with the username hash we want to use and a flag to turn it on and off? I really want to prevent some WTF moments downstream. Say sometime in the future someone makes a retired_user_joe account, and after we have all forgotten about GDPR someone sees that and decides that we need to scrub that users data because they think they slipped through the process somehow? I think we should attempt to avoid this type of thing the best we can.

Reply:
I see - your concern is someone in the future assuming that all usernames with a particular prefix are retired users. OK - how about a Django setting named `DISALLOWED_USERNAME_PREFIX` that holds a single string and that string is not allowed as a prefix of any new registrants' usernames? And if the setting is `None`, the check is skipped?

BTW - we're in luck according to the prod-edx-edxapp DB replica:
```
mysql> select id, username from auth_user where username like 'retired_user_%';
Empty set (0.00 sec)
```